pkg/server/server.go

```diff
@@ -287,12 +284,146 @@ func (h *handler) authenticateEndpoint(w http.ResponseWriter, req *http.Request)
 	})
 }
 
-func renderTemplate(template string, identity *token.Identity) string {
+func (h *handler) doMapping(identity *token.Identity, arn string) (string, []string, error) {
+	if roleMapping, exists := h.lowercaseRoleMap[arn]; exists {
+		username, err := h.renderTemplate(roleMapping.Username, identity)
+		if err != nil {
+			return "", nil, fmt.Errorf("failed mapping username: %s", err.Error())
+		}
+		groups := []string{}
+		for _, groupPattern := range roleMapping.Groups {
+			group, err := h.renderTemplate(groupPattern, identity)
+			if err != nil {
+				return "", nil, fmt.Errorf("failed mapping group: %s", err.Error())
+			}
+			groups = append(groups, group)
+		}
+		return username, groups, nil
+	} else if userMapping, exists := h.lowercaseUserMap[arn]; exists {
+		return userMapping.Username, userMapping.Groups, nil
+	} else if _, exists := h.accountMap[identity.AccountID]; exists {
+		return identity.CanonicalARN, []string{}, nil
+	}
+	return "", nil, fmt.Errorf("ARN is not mapped: %s", arn)
+}
+
+func (h *handler) renderTemplate(template string, identity *token.Identity) (string, error) {
+
+	// Private DNS requires EC2 API call
+	if strings.Contains(template, "{{EC2PrivateDNSName}}") {
+		re := regexp.MustCompile("^i-(\\w{8}|\\w{17})$")
+		match := re.MatchString(identity.SessionName)
+		if !match {
+			return "", fmt.Errorf("SessionName did not contain an instance id")
+		}
+		privateDNSName, err := h.ec2Provider.getPrivateDNSName(identity.SessionName)
+		if err != nil {
+			return "", err
+		}
+		template = strings.Replace(template, "{{EC2PrivateDNSName}}", privateDNSName, -1)
+	}
+
+	template = strings.Replace(template, "{{AccountID}}", identity.AccountID, -1)
+
 	// usernames and groups must be a DNS-1123 hostname matching the regex
 	// "[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*"
 	sessionName := strings.Replace(identity.SessionName, "@", "-", -1)
-
-	template = strings.Replace(template, "{{AccountID}}", identity.AccountID, -1)
 	template = strings.Replace(template, "{{SessionName}}", sessionName, -1)
-	return template
+
```

Comment:
Since usernames must be DNS-1123 compliant should we have a final check to ensure the rendered template is valid and return an error if it is not?  For instance if I accidentally put in `{{AccountId}}` instead of `{{AccountID}}` it will not get replaced and we will return an invalid username to k8s.